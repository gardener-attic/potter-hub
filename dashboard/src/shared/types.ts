import * as jsonSchema from "json-schema";
import { IAuthState } from "../reducers/auth";
import { IClusterBomState } from "../reducers/clusterbom";
import { IConfigState } from "../reducers/config";
import { INamespaceState } from "../reducers/namespace";
import { IAppRepositoryState } from "../reducers/repos";
import { IServiceCatalogState } from "../reducers/serviceCatalog";
import { hapi } from "./hapi/release";
import { ResourceKind } from "./ResourceKinds";

// Allow defining multiple error classes
// tslint:disable:max-classes-per-file

class CustomError extends Error {
  // The constructor is defined so we can later on compare the returned object
  // via err.contructor  == FOO
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
export class ForbiddenError extends CustomError {}
export class UnauthorizedError extends CustomError {}

export class NotFoundError extends CustomError {}

export class ConflictError extends CustomError {}

export class UnprocessableEntity extends CustomError {}

export interface IRepo {
  name: string;
  url: string;
}

export interface IChartVersion {
  id: string;
  attributes: IChartVersionAttributes;
  relationships: {
    chart: {
      data: IChartAttributes;
    };
  };
}

export interface IChartVersionAttributes {
  version: string;
  app_version: string;
  created: string;
}

export interface IChart {
  id: string;
  attributes: IChartAttributes;
  relationships: {
    latestChartVersion: {
      data: IChartVersionAttributes;
    };
  };
}

export interface IChartAttributes {
  name: string;
  description: string;
  home?: string;
  icon?: string;
  keywords: string[];
  maintainers: Array<{
    name: string;
    email?: string;
  }>;
  repo: IRepo;
  sources: string[];
}

export interface IChartState {
  isFetching: boolean;
  selected: {
    error?: Error;
    version?: IChartVersion;
    versions: IChartVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
    schema?: any;
  };
  deployed: {
    chartVersion?: IChartVersion;
    values?: string;
    schema?: jsonSchema.JSONSchema4;
  };
  items: IChart[];
}

export interface IChartUpdateInfo {
  upToDate: boolean;
  chartLatestVersion: string;
  appLatestVersion: string;
  repository: IRepo;
  error?: Error;
}

export interface IDeployment {
  metadata: {
    name: string;
    namespace: string;
  };
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip?: string; hostname?: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IHTTPIngressPath {
  path: string;
}
export interface IIngressHTTP {
  paths: IHTTPIngressPath[];
}
export interface IIngressRule {
  host: string;
  http: IIngressHTTP;
}

export interface IIngressTLS {
  hosts: string[];
}

export interface IIngressSpec {
  rules: IIngressRule[];
  tls?: IIngressTLS[];
}

export interface IResource {
  apiVersion: string;
  kind: ResourceKind;
  type: string;
  spec: any;
  status: any;
  metadata: {
    name: string;
    namespace: string;
    annotations?: { [key: string]: string };
    creationTimestamp: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    resourceVersion: string;
    uid: string;
    selfLink: string;
  };
}

export interface IOwnerReference {
  apiVersion: string;
  blockOwnerDeletion: boolean;
  kind: string;
  name: string;
  uid: string;
}

export interface ISecret {
  apiVersion: string;
  kind: string;
  type: string;
  data: { [s: string]: string };
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IDeploymentStatus {
  replicas: number;
  updatedReplicas: number;
  availableReplicas: number;
}

export interface IStatefulsetStatus {
  replicas: number;
  updatedReplicas: number;
  readyReplicas: number;
}

export interface IDaemonsetStatus {
  currentNumberScheduled: number;
  numberReady: number;
}

export interface IRelease extends hapi.release.Release {
  updateInfo?: IChartUpdateInfo;
}

export interface IAppState {
  isFetching: boolean;
  error?: Error;
  deleteError?: Error;
  // currently items are always Helm releases
  items: IRelease[];
  listingAll: boolean;
  listOverview?: IAppOverview[];
  selected?: IRelease;
}

export interface IStoreState {
  serviceCatalog: IServiceCatalogState;
  apps: IAppState;
  auth: IAuthState;
  charts: IChartState;
  clusterBom: IClusterBomState;
  config: IConfigState;
  kube: IKubeState;
  repos: IAppRepositoryState;
  namespace: INamespaceState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#objects */
export interface IK8sObject<M, SP, ST> extends IK8sResource {
  metadata: {
    annotations?: { [key: string]: string };
    creationTimestamp?: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    name: string;
    namespace: string;
    resourceVersion?: string;
    uid: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
  spec?: SP;
  status?: ST;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IAppRepository
  extends IK8sObject<
    {
      clusterName: string;
      creationTimestamp: string;
      deletionGracePeriodSeconds: string | null;
      deletionTimestamp: string | null;
      resourceVersion: string;
      selfLink: string;
    },
    {
      type: string;
      url: string;
      auth: {
        header: {
          secretKeyRef: {
            name: string;
            key: string;
          };
        };
      };
      resyncRequests: number;
    },
    undefined
  > {}

export interface IAppRepositoryList
  extends IK8sList<
    IAppRepository,
    {
      continue: string;
      resourceVersion: string;
      selfLink: string;
    }
  > {}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#response-status-kind */
export interface IStatus extends IK8sResource {
  kind: "Status";
  status: "Success" | "Failure";
  message: string;
  reason:
    | "BadRequest"
    | "Unauthorized"
    | "Forbidden"
    | "NotFound"
    | "AlreadyExists"
    | "Conflict"
    | "Invalid"
    | "Timeout"
    | "ServerTimeout"
    | "MethodNotAllowed"
    | "InternalError";
  details?: {
    kind?: string;
    name?: string;
    causes?: IStatusCause[] | string;
  };
}

interface IStatusCause {
  field: string;
  message: string;
  reason: string;
}

export interface IRouterPathname {
  router: {
    location: {
      pathname: string;
    };
  };
}

export interface IRuntimeVersion {
  name: string;
  version: string;
  runtimeImage: string;
  initImage: string;
}

export interface IRuntime {
  ID: string;
  versions: IRuntimeVersion[];
  depName: string;
  fileNameSuffix: string;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IAppOverview {
  releaseName: string;
  namespace: string;
  version: string;
  icon?: string;
  status: string;
  chart: string;
  chartMetadata: hapi.chart.Metadata;
  // UpdateInfo is internally populated
  updateInfo?: IChartUpdateInfo;
  description: string;
}

export interface IKubeItem<T> {
  isFetching: boolean;
  item?: T;
  error?: Error;
}

export interface IKubeState {
  items: { [s: string]: IKubeItem<IResource> };
  fetchTargetClusterSecretErr?: Error;
  sockets: { [s: string]: { socket: WebSocket; closeTimer: () => void } };
}

export interface IBasicFormParam {
  path: string;
  type?: jsonSchema.JSONSchema4TypeName | jsonSchema.JSONSchema4TypeName[];
  value?: any;
  title?: string;
  minimum?: number;
  maximum?: number;
  render?: string;
  description?: string;
  enum?: string[];
  hidden?:
    | {
        path: any;
        value: string;
        conditions: Array<{
          path: any;
          value: string;
        }>;
        operator: string;
      }
    | string;
  children?: IBasicFormParam[];
}
export interface IBasicFormSliderParam extends IBasicFormParam {
  sliderMin?: number;
  sliderMax?: number;
  sliderUnit?: string;
}

export interface IApplicationConfig {
  id: string;
  configType: string;
  typeSpecificData: {};
  values: {};
  secretValues: {};
  noReconcile: boolean;
  readyRequirements: IReadyRequirements;
}

export interface IReadyRequirements {
    jobs: IJobReadyRequirement[];
    resources: IResourceReadyRequirement[];
}

export interface IJobReadyRequirement {
    name: string;
    namespace: string
}

export interface IResourceReadyRequirement {
    name: string;
    namespace: string;
    apiVersion: string;
    resource: string;
    fieldPath: string;
    successValues: ISuccessValue[];
}

export interface ISuccessValue {
    value: {};
}

export interface IAutoDelete {
    clusterBomAge: number;
}

export interface IClusterBom
  extends IK8sObject<
    {},
    {
      secretRef: string;
      applicationConfigs: IApplicationConfig[];
      autoDelete: IAutoDelete;
    },
    {
      observedGeneration: number;
      overallState: string;
      overallTime: string;
      conditions: ICondition[];
      applicationStates: IApplicationState[];
      overallNumOfDeployments: number;
      overallNumOfReadyDeployments: number;
      overallProgress: number;
    }
  > {}

export interface ICondition {
  type: string;
  status: string;
  lastUpdateTime: string;
  lastTransitionTime: string;
  reason: string;
  message: string;
}

export interface IApplicationState {
  id: string;
  state: string;
  detailedState: {};
}

export interface IClusterBomList extends IK8sList<IClusterBom, {}> {}

export interface IURLParams {
    targetClusterSecretName: string;
    targetClusterSecretNamespace: string;
}
