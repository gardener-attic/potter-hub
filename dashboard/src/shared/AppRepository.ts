import { axiosWithAuth } from "./AxiosInstance";
import { UI_BACKEND_ROOT_URL } from "./Kube";
import { IAppRepository, IAppRepositoryList } from "./types";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get<IAppRepositoryList>(
      AppRepository.getResourceLink(namespace),
    );
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async update(name: string, namespace: string, newApp: IAppRepository) {
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(name, namespace), newApp);
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async create(
    name: string,
    namespace: string,
    url: string,
    auth: any,
    syncJobPodTemplate: any,
  ) {
    const { data } = await axiosWithAuth.post<IAppRepository>(
      AppRepository.getResourceLink(namespace),
      {
        apiVersion: "kubeapps.com/v1alpha1",
        kind: "AppRepository",
        metadata: {
          name,
        },
        spec: { auth, type: "helm", url, syncJobPodTemplate },
      },
    );
    return data;
  }

  private static APIBase: string = UI_BACKEND_ROOT_URL;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apprepositories`;
  
  private static getResourceLink(namespace?: string): string {
    return AppRepository.APIEndpoint;
  }

  private static getSelfLink(name: string, namespace: string): string {
    return `${AppRepository.APIEndpoint}/${name}`;
  }
}
