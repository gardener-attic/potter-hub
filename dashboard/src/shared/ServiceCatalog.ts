import { JSONSchema6 } from "json-schema";
import * as urls from "../shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { IClusterServiceClass } from "./ClusterServiceClass";
import { K8S_REVERSE_PROXY_URL } from "./Kube";
import { IServiceInstance } from "./ServiceInstance";
import { IK8sList, IStatus } from "./types";

export class ServiceCatalog {
  public static async getServiceClasses() {
    return this.getItems<IClusterServiceClass>("clusterserviceclasses");
  }

  public static async getServiceBrokers() {
    return this.getItems<IServiceBroker>("clusterservicebrokers");
  }

  public static async getServicePlans() {
    return this.getItems<IServicePlan>("clusterserviceplans");
  }

  public static async deprovisionInstance(instance: IServiceInstance) {
    const { data } = await axiosWithAuth.delete(
      `${K8S_REVERSE_PROXY_URL}${instance.metadata.selfLink}`,
    );
    return data;
  }

  public static async syncBroker(broker: IServiceBroker) {
    const { data } = await axiosWithAuth.patch<IStatus>(
      urls.api.clusterservicebrokers.sync(broker),
      {
        spec: {
          relistRequests: broker.spec.relistRequests + 1,
        },
      },
      {
        headers: { "Content-Type": "application/merge-patch+json" },
        validateStatus: statusCode => true,
      },
    );
    return data;
  }

  public static async isCatalogInstalled(): Promise<boolean> {
    try {
      const { status } = await axiosWithAuth.get(this.endpoint);
      return status === 200;
    } catch (err) {
      return false;
    }
  }

  public static async getItems<T>(resource: string, namespace?: string): Promise<T[]> {
    const response = await axiosWithAuth.get<IK8sList<T, {}>>(
      this.endpoint + (namespace ? `/namespaces/${namespace}` : "") + `/${resource}`,
    );
    const json = response.data;
    return json.items;
  }

  private static endpoint: string = `${K8S_REVERSE_PROXY_URL}/apis/servicecatalog.k8s.io/v1beta1`;
}
export interface IK8sApiListResponse<T> {
  kind: string;
  apiVersion: string;
  metadata: {
    selfLink: string;
    resourceVersion: string;
  };
  items: T[];
}

export interface ICondition {
  type: string;
  status: string;
  lastTransitionTime: string;
  reason: string;
  message: string;
}

export interface IServiceBroker {
  metadata: {
    name: string;
    selfLink: string;
    uid: string;
    resourceVersion: string;
    generation: number;
    creationTimestamp: string;
    finalizers: string[];
  };
  spec: {
    url: string;
    authInfo: any; // Look into
    relistBehavior: string;
    relistDuration: string;
    relistRequests: number;
  };
  status: {
    conditions: ICondition[];
    reconciledGeneration: number;
    lastCatalogRetrievalTime: string;
  };
}

export interface IServicePlan {
  metadata: {
    name: string;
    selfLink: string;
    uid: string;
    resourceVersion: string;
    creationTimestamp: string;
  };
  spec: {
    clusterServiceBrokerName: string;
    externalName: string;
    externalID: string;
    description: string;
    externalMetadata?: {
      displayName: string;
      bullets: string[];
    };
    instanceCreateParameterSchema?: JSONSchema6;
    serviceBindingCreateParameterSchema?: JSONSchema6;
    free: boolean;
    clusterServiceClassRef: {
      name: string;
    };
  };
  status: {
    removedFromBrokerCatalog: boolean;
  };
}
