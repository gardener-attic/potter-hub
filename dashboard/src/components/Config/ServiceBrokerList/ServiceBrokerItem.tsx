import * as React from "react";

import { CardFooter } from "../../../components/Card";
import InfoCard from "../../../components/InfoCard";
import azureIcon from "../../../icons/azure.png";
import gcpIcon from "../../../icons/gcp.png";
import placeholder from "../../../img/placeholder.png";
import { IServiceBroker } from "../../../shared/ServiceCatalog";
import SyncButton from "./SyncButton";

interface IServiceBrokerItemProps {
  broker: IServiceBroker;
  sync: (broker: IServiceBroker) => Promise<any>;
}

function getIcon(brokerURL: string) {
  const parser = document.createElement("a");
  parser.href = brokerURL;
  if (parser.hostname.match(/azure/)) {
    return azureIcon;
  } else if (parser.hostname.match(/servicebroker\.googleapis/)) {
    return gcpIcon;
  }
  return placeholder;
}

const ServiceBrokerItem: React.SFC<IServiceBrokerItemProps> = props => {
  const { broker, sync } = props;
  return (
    <InfoCard
      title={broker.metadata.name}
      info={`Last updated ${broker.status.lastCatalogRetrievalTime}`}
      icon={getIcon(broker.spec.url)}
    >
      <CardFooter className="text-c">
        <SyncButton sync={sync} broker={broker} />
      </CardFooter>
    </InfoCard>
  );
};

export default ServiceBrokerItem;
