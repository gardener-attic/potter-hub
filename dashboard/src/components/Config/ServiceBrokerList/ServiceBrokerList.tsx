import * as React from "react";

import LoadingWrapper from "../../../components/LoadingWrapper";
import PageHeader from "../../../components/PageHeader";
import { IServiceCatalogState } from "../../../reducers/serviceCatalog";
import { IServiceBroker } from "../../../shared/ServiceCatalog";
import { IRBACRole } from "../../../shared/types";
import { CardGrid } from "../../Card";
import {
  ErrorSelector,
  ServiceBrokersNotFoundAlert,
  ServiceCatalogNotInstalledAlert,
} from "../../ErrorAlert";
import ServiceBrokerItem from "./ServiceBrokerItem";

interface IServiceBrokerListProps {
  errors: {
    fetch?: Error;
    update?: Error;
  };
  getBrokers: () => Promise<any>;
  brokers: IServiceCatalogState["brokers"];
  sync: (broker: IServiceBroker) => Promise<any>;
  checkCatalogInstalled: () => Promise<any>;
  isInstalled: boolean;
}

export const RequiredRBACRoles: { [s: string]: IRBACRole[] } = {
  resync: [
    {
      apiGroup: "servicecatalog.k8s.io",
      clusterWide: true,
      resource: "clusterservicebrokers",
      verbs: ["patch"],
    },
  ],
  view: [
    {
      apiGroup: "servicecatalog.k8s.io",
      clusterWide: true,
      resource: "clusterservicebrokers",
      verbs: ["list"],
    },
  ],
};

class ServiceBrokerList extends React.Component<IServiceBrokerListProps> {
  public componentDidMount() {
    this.props.checkCatalogInstalled();
    this.props.getBrokers();
  }

  public render() {
    const { brokers, errors, sync, isInstalled } = this.props;
    let body = <span />;
    if (errors.fetch) {
      body = (
        <ErrorSelector
          error={errors.fetch}
          resource="Service Brokers"
          action="view"
          defaultRequiredRBACRoles={RequiredRBACRoles}
        />
      );
    } else {
      if (brokers.list.length > 0) {
        if (errors.update) {
          body = (
            <ErrorSelector
              error={errors.update}
              resource="Service Brokers"
              action="resync"
              defaultRequiredRBACRoles={RequiredRBACRoles}
            />
          );
        } else {
          body = (
            <CardGrid className="BrokerList">
              {brokers.list.map(broker => (
                <ServiceBrokerItem key={broker.metadata.uid} broker={broker} sync={sync} />
              ))}
            </CardGrid>
          );
        }
      } else {
        body = <ServiceBrokersNotFoundAlert />;
      }
    }
    return (
      <section className="AppList">
        <PageHeader>
          <h1>Service Brokers</h1>
        </PageHeader>
        <LoadingWrapper loaded={!brokers.isFetching}>
          {!isInstalled ? <ServiceCatalogNotInstalledAlert /> : <main>{body}</main>}
        </LoadingWrapper>
      </section>
    );
  }
}

export default ServiceBrokerList;
