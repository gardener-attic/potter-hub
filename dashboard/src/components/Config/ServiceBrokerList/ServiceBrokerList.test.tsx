import { mount, shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import ServiceBrokerList from ".";
import {
  ErrorSelector,
  ServiceBrokersNotFoundAlert,
  ServiceCatalogNotInstalledAlert,
} from "../../../components/ErrorAlert";
import { IServiceBroker } from "../../../shared/ServiceCatalog";
import itBehavesLike from "../../../shared/specs";
import { ForbiddenError } from "../../../shared/types";

let defaultProps = {
  getBrokers: jest.fn(),
  brokers: { isFetching: false, list: [] },
  sync: jest.fn(),
  errors: {},
  checkCatalogInstalled: jest.fn(),
  isInstalled: true,
};

beforeEach(() => {
  // Restart mock stats
  defaultProps = {
    getBrokers: jest.fn(),
    brokers: { isFetching: false, list: [] },
    sync: jest.fn(),
    errors: {},
    checkCatalogInstalled: jest.fn(),
    isInstalled: true,
  };
});

const broker = {
  metadata: { name: "wall-street", uid: "1" },
  spec: {
    url: "https://foo.bar",
  },
  status: {
    lastCatalogRetrievalTime: "today",
  },
} as IServiceBroker;

context("if the service broker is not installed", () => {
  it("shows a warning message", () => {
    const props = { ...defaultProps, isInstalled: false };
    const wrapper = shallow(<ServiceBrokerList {...props} />);
    expect(wrapper.find(ServiceCatalogNotInstalledAlert)).toExist();
    expect(wrapper).toMatchSnapshot();
  });
});

context("while fetching brokers", () => {
  const props = { ...defaultProps, brokers: { isFetching: true, list: [] } };

  itBehavesLike("aLoadingComponent", { component: ServiceBrokerList, props });

  it("matches the snapshot", () => {
    const wrapper = shallow(<ServiceBrokerList {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a Application header", () => {
    const wrapper = shallow(<ServiceBrokerList {...props} />);
    expect(wrapper.find("h1").text()).toContain("Service Brokers");
  });
});

context("when all the brokers are loaded", () => {
  it("shows a warning to install no service broker is installed", () => {
    const wrapper = mount(<ServiceBrokerList {...defaultProps} />);
    expect(wrapper.find(ServiceBrokersNotFoundAlert)).toExist();
    expect(wrapper).toMatchSnapshot();
  });

  it("shows a forbiden (fetch) error if it exists", () => {
    const wrapper = shallow(
      <ServiceBrokerList {...defaultProps} errors={{ fetch: new ForbiddenError() }} />,
    );
    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper).toMatchSnapshot();
  });

  it("shows a forbiden (resync) error if it exists", () => {
    const wrapper = mount(
      <ServiceBrokerList
        {...defaultProps}
        brokers={{ isFetching: false, list: [broker] }}
        errors={{ update: new ForbiddenError() }}
      />,
    );
    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper).toMatchSnapshot();
  });

  it("when there are no errors, renders the broker list", () => {
    const wrapper = shallow(
      <ServiceBrokerList {...defaultProps} brokers={{ list: [broker], isFetching: false }} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
