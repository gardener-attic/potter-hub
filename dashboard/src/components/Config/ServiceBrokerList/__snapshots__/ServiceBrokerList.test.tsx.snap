// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if the service broker is not installed shows a warning message 1`] = `
<section
  className="AppList"
>
  <PageHeader>
    <h1>
      Service Brokers
    </h1>
  </PageHeader>
  <LoadingWrapper
    loaded={true}
    type={0}
  >
    <ServiceCatalogNotInstalledAlert />
  </LoadingWrapper>
</section>
`;

exports[`when all the brokers are loaded shows a forbiden (fetch) error if it exists 1`] = `
<section
  className="AppList"
>
  <PageHeader>
    <h1>
      Service Brokers
    </h1>
  </PageHeader>
  <LoadingWrapper
    loaded={true}
    type={0}
  >
    <main>
      <ErrorSelector
        action="view"
        defaultRequiredRBACRoles={
          Object {
            "resync": Array [
              Object {
                "apiGroup": "servicecatalog.k8s.io",
                "clusterWide": true,
                "resource": "clusterservicebrokers",
                "verbs": Array [
                  "patch",
                ],
              },
            ],
            "view": Array [
              Object {
                "apiGroup": "servicecatalog.k8s.io",
                "clusterWide": true,
                "resource": "clusterservicebrokers",
                "verbs": Array [
                  "list",
                ],
              },
            ],
          }
        }
        error={[Error]}
        resource="Service Brokers"
      />
    </main>
  </LoadingWrapper>
</section>
`;

exports[`when all the brokers are loaded shows a forbiden (resync) error if it exists 1`] = `
<ServiceBrokerList
  brokers={
    Object {
      "isFetching": false,
      "list": Array [
        Object {
          "metadata": Object {
            "name": "wall-street",
            "uid": "1",
          },
          "spec": Object {
            "url": "https://foo.bar",
          },
          "status": Object {
            "lastCatalogRetrievalTime": "today",
          },
        },
      ],
    }
  }
  checkCatalogInstalled={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  errors={
    Object {
      "update": [Error],
    }
  }
  getBrokers={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  isInstalled={true}
  sync={[MockFunction]}
>
  <section
    className="AppList"
  >
    <PageHeader>
      <header
        className="PageHeader"
      >
        <div
          className="row padding-t-big padding-b-small collapse-b-phone-land align-center"
        >
          <h1>
            Service Brokers
          </h1>
        </div>
      </header>
    </PageHeader>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <main>
        <ErrorSelector
          action="resync"
          defaultRequiredRBACRoles={
            Object {
              "resync": Array [
                Object {
                  "apiGroup": "servicecatalog.k8s.io",
                  "clusterWide": true,
                  "resource": "clusterservicebrokers",
                  "verbs": Array [
                    "patch",
                  ],
                },
              ],
              "view": Array [
                Object {
                  "apiGroup": "servicecatalog.k8s.io",
                  "clusterWide": true,
                  "resource": "clusterservicebrokers",
                  "verbs": Array [
                    "list",
                  ],
                },
              ],
            }
          }
          error={[Error]}
          resource="Service Brokers"
        >
          <PermissionsErrorPage
            action="resync Service Brokers"
            namespace=""
            roles={
              Array [
                Object {
                  "apiGroup": "servicecatalog.k8s.io",
                  "clusterWide": true,
                  "resource": "clusterservicebrokers",
                  "verbs": Array [
                    "patch",
                  ],
                },
              ]
            }
          >
            <UnexpectedErrorPage
              icon={[Function]}
              showGenericMessage={false}
              title={
                <span>
                  You don't have sufficient permissions to 
                  resync Service Brokers
                   in 
                  
                </span>
              }
            >
              <div
                className="alert alert-error margin-v-bigger"
              >
                <ErrorPageHeader
                  icon={[Function]}
                >
                  <header>
                    <div
                      className="margin-b-big"
                    >
                      <h5
                        className="type-regular"
                      >
                        <span
                          className="error__icon margin-r-small"
                        >
                          <Lock
                            color="currentColor"
                            size="24"
                          >
                            <svg
                              fill="none"
                              height="24"
                              stroke="currentColor"
                              strokeLinecap="round"
                              strokeLinejoin="round"
                              strokeWidth="2"
                              viewBox="0 0 24 24"
                              width="24"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <rect
                                height="11"
                                rx="2"
                                ry="2"
                                width="18"
                                x="3"
                                y="11"
                              />
                              <path
                                d="M7 11V7a5 5 0 0 1 10 0v4"
                              />
                            </svg>
                          </Lock>
                        </span>
                        <span>
                          You don't have sufficient permissions to 
                          resync Service Brokers
                           in 
                        </span>
                      </h5>
                    </div>
                  </header>
                </ErrorPageHeader>
                <div
                  className="error__content margin-l-enormous"
                >
                  <div>
                    <p>
                      Ask your administrator for the following RBAC roles:
                    </p>
                    <ul
                      className="error__permisions-list"
                    >
                      <PermissionsListItem
                        key="0"
                        namespace=""
                        role={
                          Object {
                            "apiGroup": "servicecatalog.k8s.io",
                            "clusterWide": true,
                            "resource": "clusterservicebrokers",
                            "verbs": Array [
                              "patch",
                            ],
                          }
                        }
                      >
                        <li>
                          patch
                           
                          <code>
                            clusterservicebrokers
                          </code>
                           
                          <span>
                            (
                            <code>
                              servicecatalog.k8s.io
                            </code>
                            )
                          </span>
                           
                          in 
                          all namespaces
                          .
                        </li>
                      </PermissionsListItem>
                    </ul>
                    <p>
                      Ask you Gardener project manager to add you as a member, so that you have access to the cluster.
                    </p>
                  </div>
                </div>
              </div>
            </UnexpectedErrorPage>
          </PermissionsErrorPage>
        </ErrorSelector>
      </main>
    </LoadingWrapper>
  </section>
</ServiceBrokerList>
`;

exports[`when all the brokers are loaded shows a warning to install no service broker is installed 1`] = `
<ServiceBrokerList
  brokers={
    Object {
      "isFetching": false,
      "list": Array [],
    }
  }
  checkCatalogInstalled={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  errors={Object {}}
  getBrokers={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  isInstalled={true}
  sync={[MockFunction]}
>
  <section
    className="AppList"
  >
    <PageHeader>
      <header
        className="PageHeader"
      >
        <div
          className="row padding-t-big padding-b-small collapse-b-phone-land align-center"
        >
          <h1>
            Service Brokers
          </h1>
        </div>
      </header>
    </PageHeader>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <main>
        <ServiceBrokersNotFoundAlert>
          <NotFoundErrorPage
            header="No Service Brokers installed."
          >
            <div
              className="alert alert-error margin-t-bigger"
            >
              <ErrorPageHeader
                icon={[Function]}
              >
                <header>
                  <div
                    className="margin-b-big"
                  >
                    <h5
                      className="type-regular"
                    >
                      <span
                        className="error__icon margin-r-small"
                      >
                        <X
                          color="currentColor"
                          size="24"
                        >
                          <svg
                            fill="none"
                            height="24"
                            stroke="currentColor"
                            strokeLinecap="round"
                            strokeLinejoin="round"
                            strokeWidth="2"
                            viewBox="0 0 24 24"
                            width="24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <line
                              x1="18"
                              x2="6"
                              y1="6"
                              y2="18"
                            />
                            <line
                              x1="6"
                              x2="18"
                              y1="6"
                              y2="18"
                            />
                          </svg>
                        </X>
                      </span>
                      No Service Brokers installed.
                    </h5>
                  </div>
                </header>
              </ErrorPageHeader>
              <div
                className="error__content margin-l-enormous"
              >
                <p>
                  Ask an administrator to install a compatible
                   
                  <a
                    href="https://github.com/osbkit/brokerlist"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Service Broker
                  </a>
                   
                  to browse, provision and manage external services within Kubeapps.
                </p>
              </div>
            </div>
          </NotFoundErrorPage>
        </ServiceBrokersNotFoundAlert>
      </main>
    </LoadingWrapper>
  </section>
</ServiceBrokerList>
`;

exports[`when all the brokers are loaded when there are no errors, renders the broker list 1`] = `
<section
  className="AppList"
>
  <PageHeader>
    <h1>
      Service Brokers
    </h1>
  </PageHeader>
  <LoadingWrapper
    loaded={true}
    type={0}
  >
    <main>
      <CardGrid
        className="BrokerList"
      >
        <ServiceBrokerItem
          broker={
            Object {
              "metadata": Object {
                "name": "wall-street",
                "uid": "1",
              },
              "spec": Object {
                "url": "https://foo.bar",
              },
              "status": Object {
                "lastCatalogRetrievalTime": "today",
              },
            }
          }
          key="1"
          sync={[MockFunction]}
        />
      </CardGrid>
    </main>
  </LoadingWrapper>
</section>
`;

exports[`while fetching brokers loading spinner matches the snapshot 1`] = `
<section
  className="AppList"
>
  <PageHeader>
    <h1>
      Service Brokers
    </h1>
  </PageHeader>
  <LoadingWrapper
    loaded={false}
    type={0}
  >
    <main>
      <ServiceBrokersNotFoundAlert />
    </main>
  </LoadingWrapper>
</section>
`;

exports[`while fetching brokers matches the snapshot 1`] = `
<section
  className="AppList"
>
  <PageHeader>
    <h1>
      Service Brokers
    </h1>
  </PageHeader>
  <LoadingWrapper
    loaded={false}
    type={0}
  >
    <main>
      <ServiceBrokersNotFoundAlert />
    </main>
  </LoadingWrapper>
</section>
`;
